lib/virt_autotest/utils.pm

```diff
@@ -439,31 +439,22 @@ sub wait_guest_online {
     }
 }
 
-# Shutdown all guests and wait until they are shutdown
+# Shutdown all guests
 sub shutdown_guests {
-    ## Reboot the guest to ensure the settings are applied
-    # Shutdown and start the guest because some might have the on_reboot=destroy policy still applied
-    script_run("virsh shutdown $_") foreach (keys %virt_autotest::common::guests);
-    # Wait until guests are terminated
-    wait_guests_shutdown();
-}
-
-# wait_guests_shutdown([$timeout]) waits for all guests to be shutdown
-sub wait_guests_shutdown {
-    my $retries = shift // 240;
-    # Note: Domain-0 is for xen only, but it does not hurt to exclude this also in kvm runs.
-    # Firstly wait for guest shutdown for a while, turn it off forcibly using "virsh destroy" if timed-out.
-    # Then wait for guest shutdown again with default "die => 1".
-    if (script_retry("! virsh list | grep -v Domain-0 | grep running", timeout => 60, delay => 1, retry => $retries, die => 0) != 0) {
-        script_run("virsh destroy $_") foreach (keys %virt_autotest::common::guests);
+    my @guests = @_;
+    foreach my $guest (@guests) {
+        script_run("virsh shutdown $guest");
+        script_retry("virsh domstate $guest|grep 'shut off'", retry => 5);    #assert if vm fails shutdown
     }
```

Comment:
I think some tests calling this shutdown_guests or wait_guests_shutdown might just need guest in "shut off" regardless of whether it is turned off gracefully or not. You removed "virsh destroy", so some tests may fail ? Have you covered enough tests ?

Reply:
I'd just like to know if there could be any bug that makes shutdown fail. This change is out of scope this ticket.  If you want "vrish destroy", I can add it.

Reply:
I think adding "virsh destroy" is better at the moment. Actually "virsh shutdown" is not any magic thing, it just sends a signal to virtual machine. And virtual machine can ignore such signal under certain circumstances which depends on configuration and operating system. I do not think this is a buggy area. I have ever encountered such issue, the OS just popped up a dialog to  let user know and confirm, but there is no user logged in, so it never shuts down. You can refer to this [article1](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/virtualization_deployment_and_administration_guide/sect-managing_guest_virtual_machines_with_virsh-shutting_down_rebooting_and_force_shutdown_of_a_guest_virtual_machine#sect-Shutting_down_rebooting_and_force_shutdown_of_a_guest_virtual_machine-Shut_down_a_guest_virtual_machine) and [article2](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/virtualization_deployment_and_administration_guide/sect-qemu-agent-vish-shutdown). So adding "virsh destroy" can help make our lives easier.